ath79: add support to TrendNet TEW-673GRU

DeadEnded opened the pull request:
Add support for the TrendNet TEW-673GRU to ath79.
This device was supported in 19.07.9 but was deprecated with ar71xx. 
This is mostly a copy of D-Link DIR-825 B1.
Currently only sysupgrade, not factory build.
This is due to the D-Link DIR-825 B1 only supporting sysupgrade.

## Specifications

|  | |
| --- | --- |
| Architecture |	MIPS |
| Vendor |	Qualcomm Atheros |
| bootloader |	U-Boot |
| System-On-Chip |	AR7161 rev 2 (MIPS 24Kc V7.4) |
| CPU/Speed |	24Kc V7.4 680 MHz |
| Flash-Chip |	Macronix MX25L6405D |
| Flash size |	8192 KiB |
| RAM Chip:	|	ProMOS V58C2256164SCI5 × 2 |
| RAM size |	64 MiB |
| Wireless |	2 x Atheros AR922X 2.4GHz/5.0GHz 802.11abgn |
| Ethernet |	RealTek RTL8366S Gigabit w/ port based vlan support |
| USB |	Yes 2 x 2.0 |

## Flashing Instructions

Currently 21.02 is only supported through sysupgrade from 19.07.

Process:
	1) Download 19.07 tew-673gru factory bin
	2) Flash 19.07 using TrendNet GUI
	3) After device has been flashed with 19.07, download 21.02 sysupgrade.bin
	4) Flash 21.02 using OpenWRT GUI

Signed-off-by: Korey Caro <korey.caro@gmail.com>

mans0n commented:
Hi, please stick to the original PR next time.
https://openwrt.org/docs/guide-developer/frequent-pr-mistakes#do_not_multi-post_or_re-post_prs

---

And this has some formal issues, especially:
* Please drop the `(#3)` in the commit title.
* Please add device specification and flashing instruction in the commit message.
  You can copy one from the older ar71xx support commit (or the DIR-825 B1 support commit).
* Please wrap the commit message up to 75 characters per line.
* The SoB address should be surrounded by angle brackets.

https://openwrt.org/docs/guide-developer/frequent-pr-mistakes

---

The content also needs to be refined. (alphabetical sort, node/label name, indentation, ...)

Note to myself: find a good consistent form of "trendnet".

DeadEnded commented:
Firstly, my sincere apologies - I am not very experienced using git and don't know how to do many things.  I didn't realize I could re-do a commit in a PR - I should have asked.

Next, I am working through the above request, and I believe I understand everything except for the last item (`The SoB address should be surrounded by angle brackets.`).  I have looked through this commit, and other PR's and I have not been able to figure out what I need to fix for this.  Can you please clarify what I need to fix for this?

Lastly, I wanted to make sure I follow the correct process for this change (as this will be the first time I do something like this).  I do a git amend and then a force push, is that the correct process?  If not, if you could please direct me to the correct process I would be very appreciative.

Thanks!
DeadEnd

mans0n commented:
> Firstly, my sincere apologies - I am not very experienced using git and don't know how to do many things. I didn't realize I could re-do a commit in a PR - I should have asked.

Don't worry, it's expected for first-time contributors and it's totally okay. :)

> Next, I am working through the above request, and I believe I understand everything except for the last item (The SoB address should be surrounded by angle brackets.). I have looked through this commit, and other PR's and I have not been able to figure out what I need to fix for this. Can you please clarify what I need to fix for this?

I mean `Signed-off-by: name <mail-address>`.
And the commit author line should also be updated to match Signed-off-by tag.

> Lastly, I wanted to make sure I follow the correct process for this change (as this will be the first time I do something like this). I do a git amend and then a force push, is that the correct process? If not, if you could please direct me to the correct process I would be very appreciative.

Precisely. And please do not create a merge commit.

DeadEnded commented:
> Don't worry, it's expected for first-time contributors and it's totally okay. :)

Thanks for that - makes me feel less the fool.

> I mean `Signed-off-by: name <mail-address>`.
> And the commit author line should also be updated to match Signed-off-by tag.

... and then it it's so perfectly clear - I feel a fool again, lol.
I'll fix both - thanks.

> Precisely. And please do not create a merge commit.

Okay - so this one is why I'm actually replying.  I did some quick learning on the different types of commits.  If I understand correctly the issue is that other commits have been created/applied since I started this branch - so it is not current with master.  To prevent it from creating a merge commit I need to rebase from master, and then do the force push to fix my commit.  Do I have this correct?

Originally (and probably the source of my problems) I started on github and just used the webpage to fork, branch, make my changes (which led to 4 commits instead of 1 - not sure if you can do it differently on web) and then create the PR.  When I tried to fix the 4 commits to 1 commit I continued using the github site and did a squash to a temporary branch thinking that way my PR would be a single commit.  Now I'm actually using git instead of being lazy and using github pages... so that should help.

Anyway - I explained all this to ask - can I do the above rebase from master and then force push the amended commit, or has my messing around using github instead of cloning a local repo gotten me into a situation where I just need to wipe my changes on this branch and start over?

Thanks for all your help and clarifications.
I appreciate your time correcting my mistake!

DeadEnd

mans0n commented:
> To prevent it from creating a merge commit I need to rebase from master, and then do the force push to fix my commit. Do I have this correct?

Correct.

> Anyway - I explained all this to ask - can I do the above rebase from master and then force push the amended commit, or has my messing around using github instead of cloning a local repo gotten me into a situation where I just need to wipe my changes on this branch and start over?

Yes, using git CLI to rebase and force push will do the job.

Many novice contributors stumble over this git-related stuff and get frustrated, I hope you can cope with this soon. I think you are a fast learner, I believe you will be able to manage it. :)

DeadEnded commented:
Okay!
Let's see how close I got to doing it right 😄.
Let me know if you find any other mistakes or changes needed.

Thanks again!
DeadEnd

mans0n commented:
You should also change the author line, please do:
```
git config user.name <name>
git config user.email <email>
git commit --amend --reset-author
```

And could you align the specification table in the commit message so that we can read it easily?

DeadEnded committed:
ath79: add support to TrendNet TEW-673GRU

Add support for the TrendNet TEW-673GRU to ath79.
This device was supported in 19.07.9 but was deprecated with ar71xx.
This is mostly a copy of D-Link DIR-825 B1.
Updates have been completed to enable factory.bin and sysupgrade.bin both.
Code improvements to DTS file and makefile.

Architecture   |  MIPS
Vendor         |  Qualcomm Atheros
bootloader     |  U-Boot
System-On-Chip |  AR7161 rev 2 (MIPS 24Kc V7.4)
CPU/Speed      |  24Kc V7.4 680 MHz
Flash-Chip     |  Macronix MX25L6405D
Flash size     |  8192 KiB
RAM Chip:      |  ProMOS V58C2256164SCI5 × 2
RAM size       |  64 MiB
Wireless       |  2 x Atheros AR922X 2.4GHz/5.0GHz 802.11abgn
Ethernet       |  RealTek RTL8366S Gigabit w/ port based vlan support
USB            |  Yes 2 x 2.0

Initial Flashing Process:
	1) Download 22.03 tew-673gru factory bin
	2) Flash 22.03 using TrendNet GUI

OpenWRT Upgrade Process
	3) Download 22.03 tew-673gru sysupgrade.bin
	4) Flash 22.03 using OpenWRT GUI

Signed-off-by: Korey Caro <korey.caro@gmail.com>

DeadEnded commented:
Done... I hope 😄 .
I did the name and email changes you showed, but I don't know if they took.
I ran `gt config --list` and it showed both the old and the new values.. which I thought was kind of odd...
Looking on here I still see my handle instead of my name... but maybe it's correct where it matters?

Thanks!
DeadEnd

DeadEnded commented:
@mans0n I have rebased again as it has been a month since last activity.
If it's not a too much of a bother, can you please let me know if there is anything I have missed?

Thank you again for all your help!
DeadEnd

DeadEnded commented:
Rebased and force pushed again.
Bumping this as it has been 3 months since the last request.
If there is anything missing or needing fixed please let me know.

Otherwise, I am just waiting for review to be merge - thanks!
DeadEnd

stintel commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -190,6 +190,7 @@ ath79_setup_interfaces()
 		;;
 	buffalo,wzr-hp-g300nh-rb|\
 	buffalo,wzr-hp-g300nh-s|\
+	trendnet,tew-673gru|\
```
Please maintain alphabetical order

DeadEnded replied:
Thanks for catching that (and apologies you needed to).

stintel commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -599,6 +600,7 @@ ath79_setup_macs()
 	dlink,dap-3662-a1)
 		label_mac=$(mtd_get_mac_ascii bdcfg "wlanmac")
 		;;
+	trendnet,tew-673gru|\
```
Please maintain alphabetical order.

DeadEnded replied:
Thanks for catching that (and apologies you needed to).

stintel commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -153,6 +153,7 @@ case "$FIRMWARE" in
 	netgear,wndap360)
 		caldata_extract "art" 0x1000 0xeb8
 		;;
+	trendnet,tew-673gru|\
```
Please maintain alphabetical order.

DeadEnded replied:
Thanks for catching that (and apologies you needed to).

stintel commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -172,6 +173,7 @@ case "$FIRMWARE" in
 	netgear,wndap360)
 		caldata_extract "art" 0x5000 0xeb8
 		;;
+	trendnet,tew-673gru|\
```
Please maintain alphabetical order.

DeadEnded replied:
Thanks for catching that (and apologies you needed to).

stintel replied:
No need to apologize, thanks for addressing it.

stintel reviewed (changes requested):
Did not review the DTS as I don't feel qualified for that.

DragonBluep commented:
Can you try converting mac and eeprom to nvmem format? Need to test, I'm not sure if it will work.https://github.com/openwrt/openwrt/commit/251ecfe37914919422d74dd32fab15a86f65e879

DeadEnded commented:
Thanks for catching that (and apologies you needed to).


> Can you try converting mac and eeprom to nvmem format? Need to test, I'm not sure if it will work.[251ecfe](https://github.com/openwrt/openwrt/commit/251ecfe37914919422d74dd32fab15a86f65e879)

Sorry for my ignorance - can you please confirm to me what you are looking at?
I looked at some examples in other commits to see what was done...
I'm guessing you are meaning lines 605 and 606 in the `02_network` file, and/or lines 158-159 and 178-179 in `10-ath9k-eeprom`.

Either way, I will have to research how to make this conversion as I have no knowledge of how to accomplish it.  Those lines existed from the D-Link that I mimicked.

DragonBluep commented:
> I'm guessing you are meaning lines 605 and 606 in the `02_network` file, and/or lines 158-159 and 178-179 in `10-ath9k-eeprom`.
> It could also be part of the `DTS` file - but wasn't sure which parts you were looking at.

Yes, can you try this patch? Maybe WiFi is not working.

<details>
<summary>patch</summary>

```
diff --git a/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts b/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
index bce7ab9af2..4614ce9390 100644
--- a/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
+++ b/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
@@ -10,6 +10,7 @@
 	model = "TRENDNET TEW-673GRU";
 
 	aliases {
+		label-mac-device = &eth1;
 		led-boot = &wps;
 		led-failsafe = &wps;
 		led-running = &wps;
@@ -75,8 +76,6 @@
 			#size-cells = <0>;
 			status = "okay";
 
-			phy-mask = <0x10>;
-
 			phy4: ethernet-phy@4 {
 				reg = <4>;
 				phy-mode = "rgmii";
@@ -120,6 +119,9 @@
 		qca,no-eeprom;
 		#gpio-cells = <2>;
 		gpio-controller;
+
+		nvmem-cells = <&macaddr_caldata_ffa0>, <&calibration_ath9k0>;
+		nvmem-cell-names = "mac-address-ascii", "calibration";
 	};
 
 	ath9k1: wifi@0,12 {
@@ -128,6 +130,10 @@
 		qca,no-eeprom;
 		#gpio-cells = <2>;
 		gpio-controller;
+
+		nvmem-cells = <&macaddr_caldata_ffb4>, <&calibration_ath9k1>;
+		nvmem-cell-names = "mac-address-ascii", "calibration";
+		mac-address-increment = <1>;
 	};
 };
 
@@ -170,6 +176,26 @@
 				label = "caldata";
 				reg = <0x660000 0x010000>;
 				read-only;
+
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				calibration_ath9k0: calibration@1000 {
+					reg = <0x1000 0xeb8>;
+				};
+
+				calibration_ath9k1: calibration@5000 {
+					reg = <0x5000 0xeb8>;
+				};
+
+				macaddr_caldata_ffa0: macaddr@ffa0 {
+					reg = <0xffa0 0x11>;
+				};
+
+				macaddr_caldata_ffb4: macaddr@ffb4 {
+					reg = <0xffb4 0x11>;
+				};
 			};
 
 			partition@670000 {
@@ -190,6 +216,9 @@
 		speed = <1000>;
 		full-duplex;
 	};
+
+	nvmem-cells = <&macaddr_caldata_ffa0>;
+	nvmem-cell-names = "mac-address-ascii";
 };
 
 &eth1 {
@@ -198,5 +227,8 @@
 	pll-data = <0x11110000 0x00001099 0x00991099>;
 
 	phy-handle = <&phy4>;
+
+	nvmem-cells = <&macaddr_caldata_ffb4>;
+	nvmem-cell-names = "mac-address-ascii";
 };
 
diff --git a/target/linux/ath79/generic/base-files/etc/board.d/02_network b/target/linux/ath79/generic/base-files/etc/board.d/02_network
index e6b38e5696..20dc564117 100644
--- a/target/linux/ath79/generic/base-files/etc/board.d/02_network
+++ b/target/linux/ath79/generic/base-files/etc/board.d/02_network
@@ -190,8 +190,7 @@ ath79_setup_interfaces()
 		;;
 	buffalo,wzr-hp-g300nh-rb|\
 	buffalo,wzr-hp-g300nh-s|\
-	dlink,dir-825-b1|\
-	trendnet,tew-673gru)
+	dlink,dir-825-b1)
 		ucidef_set_interface_wan "eth1"
 		ucidef_add_switch "switch0" \
 			"0:lan" "1:lan" "2:lan" "3:lan" "5@eth0"
@@ -600,8 +599,7 @@ ath79_setup_macs()
 	dlink,dap-3662-a1)
 		label_mac=$(mtd_get_mac_ascii bdcfg "wlanmac")
 		;;
-	dlink,dir-825-b1|\
-	trendnet,tew-673gru)
+	dlink,dir-825-b1)
 		lan_mac=$(mtd_get_mac_text "caldata" 0xffa0)
 		wan_mac=$(mtd_get_mac_text "caldata" 0xffb4)
 		label_mac=$wan_mac
diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
index f623c3599f..df293c44db 100644
--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
+++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
@@ -153,8 +153,7 @@ case "$FIRMWARE" in
 	netgear,wndap360)
 		caldata_extract "art" 0x1000 0xeb8
 		;;
-	dlink,dir-825-b1|\
-	trendnet,tew-673gru)
+	dlink,dir-825-b1)
 		caldata_extract "caldata" 0x1000 0xeb8
 		ath9k_patch_mac_crc $(mtd_get_mac_text "caldata" 0xffa0) 0x20c
 		;;
@@ -173,8 +172,7 @@ case "$FIRMWARE" in
 	netgear,wndap360)
 		caldata_extract "art" 0x5000 0xeb8
 		;;
-	dlink,dir-825-b1|\
-	trendnet,tew-673gru)
+	dlink,dir-825-b1)
 		caldata_extract "caldata" 0x5000 0xeb8
 		ath9k_patch_mac_crc $(macaddr_add $(mtd_get_mac_text "caldata" 0xffb4) 1) 0x20c
 		;;
```
</details>

DeadEnded commented:
Currently the device running this is in use and I can't work on it until I have some downtime to take it out.

DeadEnded commented:
Just testing the additional patch that DragonBluep requested.
Unfortunately I use this device daily now and am not able to test on it.

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,202 @@
+	leds {
+		compatible = "gpio-leds";
+
+		wps: wps {
```
It's better to rename it to `led_wps: wps`

DragonBluep replied:
And also please do not forget to change `wps` to `led_wps` in `aliases` node.

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,202 @@
+			#size-cells = <0>;
+			status = "okay";
+
+			phy-mask = <0x10>;
```
Please drop `phy-mask = <0x10>;`, we do not need it anymore. https://github.com/openwrt/openwrt/commit/dae392752761128e13ede484217d806d643cbf2f

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,202 @@
+	usb_ohci_port: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
+	};
```
Please drop these led trigger:
```
	#address-cells = <1>;
	#size-cells = <0>;

	usb_ohci_port: port@1 {
		reg = <1>;
		#trigger-source-cells = <0>;
	};
```

DeadEnded replied:
to clarify, this section should now be:
```
&usb1 {
	status = "okay";
};

&usb2 {
	status = "okay";
};
```
Or do you mean that the entire section should be gone (as in no `&usb#` at all)?

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,202 @@
+
+	usb_ehci_port: port@1 {
+		reg = <1>;
+		#trigger-source-cells = <0>;
```
Please drop these led trigger:
```
	#address-cells = <1>;
	#size-cells = <0>;

	usb_ehci_port: port@1 {
		reg = <1>;
		#trigger-source-cells = <0>;
	};
```

DragonBluep commented on target/linux/ath79/image/generic.mk:
```diff
@@ -2539,6 +2539,19 @@ define Device/teltonika_rut955-h7v3c0
+  IMAGE_SIZE := 6208k
+  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | \
+		check-size | append-metadata
+  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 kmod-owl-loader
```
need to add package `kmod-switch-rtl8366s` here https://github.com/openwrt/openwrt/commit/6e0f0eae5b48bddead035d9dd89b710c10c50337

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,202 @@
+			partition@670000 {
+				label = "unknown";
+				reg = <0x670000 0x190000>;
+				read-only;
```
It's better to reuse this free space because OpenWrt image size will reach to 6 MiB soon. https://github.com/openwrt/openwrt/pull/4603

DeadEnded replied:
So remove this partition, or change the `read-only` to something else?

DragonBluep replied:
Yes, just copy all changes from the referenced PR.

DragonBluep commented:
@DeadEnded Can you backup and upload `unknow`, `caldata` partitions here? It seems to be quite different from dir-825.

And this factory image should work on your TEW-673GRU:
```
# limit the factory image size as 6144k
  IMAGE_SIZE := 6144k
  IMAGES += factory.bin
  IMAGE/factory.bin = append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
	pad-rootfs | check-size | pad-to $$$$(IMAGE_SIZE) | append-string AP94-AR7161-RT-080619-01

```

DeadEnded commented:
> 
[mtdblocks.zip](https://github.com/openwrt/openwrt/files/9880891/mtdblocks.zip)

Attached are the two partitions requested (I hope I did it correctly - saved mtdblock from GUI).

For the factory image part, do you want me to replace the below with what you sent (and remove image default and sysupgrade.bin lines)?
```
  IMAGE_SIZE := 15296k
  IMAGES := factory.bin sysupgrade.bin
  IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
	pad-rootfs
  IMAGE/factory.bin := $$(IMAGE/default) | pad-offset $$$$(IMAGE_SIZE) 26 | \
	append-string 00AP135AR9558-RT-131129-00 | check-size
  IMAGE/sysupgrade.bin := $$(IMAGE/default) | check-size | append-metadata
  ```

DragonBluep commented:
@DeadEnded The final 64k is a backup of `caldata` so we'd better don't touch it. 
<details>
<summary>You can try this patch:</summary>

```
diff --git a/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts b/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
index 27170f6eec..3270e57006 100644
--- a/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
+++ b/target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts
@@ -10,10 +10,10 @@
 	model = "TRENDNET TEW-673GRU";
 
 	aliases {
-		led-boot = &wps;
-		led-failsafe = &wps;
-		led-running = &wps;
-		led-upgrade = &wps;
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
 	};
 
 	extosc: ref {
@@ -26,28 +26,12 @@
 	leds {
 		compatible = "gpio-leds";
 
-		led_wps: wps {
-			label = "blue:wps";
+		led_power: power {
+			label = "blue:power";
 			gpios = <&gpio 9 GPIO_ACTIVE_LOW>;
 		};
 	};
 
-	ath9k-leds {
-		compatible = "gpio-leds";
-
-		wlan2g {
-			label = "blue:wlan2g";
-			gpios = <&ath9k0 5 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "phy0tpt";
-		};
-
-		wlan5g {
-			label = "blue:wlan5g";
-			gpios = <&ath9k1 5 GPIO_ACTIVE_LOW>;
-			linux,default-trigger = "phy1tpt";
-		};
-	};
-
 	keys {
 		compatible = "gpio-keys";
 
@@ -115,20 +99,16 @@
 &pcie0 {
 	status = "okay";
 
-	ath9k0: wifi@0,11 {
+	wifi@0,11 {
 		compatible = "pci168c,0029";
 		reg = <0x8800 0 0 0 0>;
 		qca,no-eeprom;
-		#gpio-cells = <2>;
-		gpio-controller;
 	};
 
-	ath9k1: wifi@0,12 {
+	wifi@0,12 {
 		compatible = "pci168c,0029";
 		reg = <0x9000 0 0 0 0>;
 		qca,no-eeprom;
-		#gpio-cells = <2>;
-		gpio-controller;
 	};
 };
 
@@ -174,7 +154,13 @@
 
 			fwconcat1: partition@670000 {
 				label = "fwconcat1";
-				reg = <0x670000 0x190000>;
+				reg = <0x670000 0x180000>;
+			};
+
+			partition@7f0000 {
+				label = "art";
+				reg = <0x7f0000 0x010000>;
+				read-only;
 			};
 		};
 	};
@@ -198,4 +184,3 @@
 
 	phy-handle = <&phy4>;
 };
-
diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
index 7bd9acae9c..423afd1717 100644
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -2559,12 +2559,13 @@ define Device/trendnet_tew-673gru
   DEVICE_VENDOR := Trendnet
   DEVICE_MODEL := TEW-673GRU
   DEVICE_VARIANT := v1.0R
-  IMAGE_SIZE := 6208k
-  IMAGE/sysupgrade.bin := append-kernel | append-rootfs | pad-rootfs | \
-  check-size | append-metadata
-  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 kmod-owl-loader | \
-  kmod-switch-rtl8366s
-  SUPPORTED_DEVICES += tew-673gru
+  # 7744k can be used but limit the factory image size as 6144k
+  IMAGE_SIZE := 6144k
+  IMAGES += factory.bin
+  IMAGE/factory.bin = append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
+	pad-rootfs | check-size | pad-to $$$$(IMAGE_SIZE) | \
+	append-string AP94-AR7161-RT-080619-01
+  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 kmod-owl-loader kmod-switch-rtl8366s
 endef
 TARGET_DEVICES += trendnet_tew-673gru
```
</details>

DeadEnded commented:
@DragonBluep done.
Just confirming - you intentionally removed the `SUPPORTED_DEVICES += tew-673gru` from generic.mk?
I see some have it and some don't - but just wanted to confirm.
Everything else should be updated.

Thank you very much for all your help!

DragonBluep commented:
> Just confirming - you intentionally removed the `SUPPORTED_DEVICES += tew-673gru` from generic.mk?
> I see some have it and some don't - but just wanted to confirm.

This is used to support the upgrade from the old version openwrt 19.07, but since we have changed the partition table, it is better not to be compatible with old version (but we can still apply it by force update). In addition, it is better to squash the three commits into one. And it is necessary to confirm whether the firmware can work normally.

DeadEnded commented:
Okay - if this is looking good, I can see about combining the commits (I have before locally used `git reset HEAD~#`) is there a better/smarter/faster way to do this?

And yes, it looks like there is only 1 LED and it's for power.

DragonBluep commented:
You can use `git rebase -i upstream/master` and then mark last two commits as squash.
And please keep this note in makefile. If one day firmware size reach to 6144k, we can change `IMAGE_SIZE` to 7744k and disable factory image to extend the life of this device.
`# max 7744k can be used but limit the factory image size as 6144k`
It's better to add these informations to commit message.

DeadEnded commented:
Okay!
I think I got it... made a few mistakes and had to fight to figure out how to get it all in correctly... but I think I got it (learning curve for a newbie).  I will try to get this built and installed as a test this next week when I can afford to have it broken 😄 just in case.

Thank you for all your help! (and hand holding)

DeadEnded commented:
Okay, I used basically all defaults (can't remember if I changed anything when I made my first (and only) build initially... but it looks like these are ready for me to test!

I know `sysupgrade.bin` I can just flash from GUI to test... what is the "correct" way to test the `factory.bin`?
Do I need to flash it to stock firmware first to test that, or is there another method? 
Not sure if you can just do a `factory.bin` from OpenWRT GUI.... never tried it 😆 

![image](https://user-images.githubusercontent.com/45110141/198419775-2eb0c218-3b52-4068-bada-bc46bd98d4f6.png)

DragonBluep commented:
You need to back to stock firmware first. You can do it by force update stock firmware in openwrt web ui.

DragonBluep commented:
Little improvement for Makefile:
```
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -2559,13 +2559,14 @@ define Device/trendnet_tew-673gru
   DEVICE_VENDOR := Trendnet
   DEVICE_MODEL := TEW-673GRU
   DEVICE_VARIANT := v1.0R
-  # max 7744k can be used but limit the factory image size as 6144k
-  IMAGE_SIZE := 6144k
+  DEVICE_PACKAGES := -uboot-envtools kmod-usb-ohci kmod-usb2 \
+       kmod-owl-loader kmod-switch-rtl8366s
+  IMAGE_SIZE := 7744k
+  FACTORY_SIZE := 6144k
   IMAGES += factory.bin
   IMAGE/factory.bin = append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
-  pad-rootfs | check-size | pad-to $$$$(IMAGE_SIZE) | \
-  append-string AP94-AR7161-RT-080619-01
-  DEVICE_PACKAGES := kmod-usb-ohci kmod-usb2 kmod-owl-loader kmod-switch-rtl8366s
+       pad-rootfs | check-size $$$$(FACTORY_SIZE) | pad-to $$$$(FACTORY_SIZE) | \
+       append-string AP94-AR7161-RT-080619-01
 endef
 TARGET_DEVICES += trendnet_tew-673gru
```

DeadEnded commented:
Update makefile.
I'll see about getting these rebuilt tomorrow and maybe tested.
Thanks!

DragonBluep commented on target/linux/ath79/dts/ar7161_trendnet_tew-673gru.dts:
```diff
@@ -0,0 +1,187 @@
+
+	phy-handle = <&phy4>;
+};
+
```
drop new line 187 here.

DragonBluep commented on target/linux/ath79/image/generic.mk:
```diff
@@ -2554,6 +2554,22 @@ define Device/teltonika_rut955-h7v3c0
+  IMAGE/factory.bin = append-kernel | pad-to $$$$(BLOCKSIZE) | append-rootfs | \
+  pad-rootfs | check-size $$$$(FACTORY_SIZE) | pad-to $$$$(FACTORY_SIZE) | \
+  append-string AP94-AR7161-RT-080619-01
+endef
```
Please use tab instead of spaces in line 2563, 2568 and 2569.

DeadEnded replied:
Ah, my editor must be messing up (aka setup wrong... user error) - I looked through and literally none of the other lines in the whole generic.mk are indented like I see on github.  I put in a tab - if it doesn't show up correctly, I'll go back and figure out how to fix it.  Thanks for catching that.

DragonBluep replied:
seems still spaces, just copy one from line 2547 and paste it here. And please add one new line in the end of the dts file. It seems that you dropped two blank lines before.

DeadEnded replied:
okay... that was comically bad... but should be fixed now.
Had to turn off "detect indentation" for makefile in VSCode... all better.
Of course then it was misaligned, but I think I have it fixed now...
Copying a tab wouldn't work because the editor wasn't honoring them... all looked like spaces or nothing at all.
Should be correct now... Sorry about all the force pushes... 😩

DragonBluep replied:
> And please add one new line in the end of the dts file. It seems that you dropped two blank lines before.

So do not forget that😄. 
And please reorganize your commit message with `git commit --amend` and delete the useless messages after ` signed-off-by xxx`.

DragonBluep reviewed (commented):
Please update firmware installation and include some mac address information in commit message.  Hope this is the final change. 949e8ba521ed7b7fd5b02e34e6706b02c17d0ae3
You can use `git commit --amend` to edit commit message.

DragonBluep reviewed (approved):
We still have ~250 KiB free space in the factory image. It may be possible to support generate a factory image under OpenWrt 23.x, but it is better to check-pick and backport to the 22.03 branch. @DeadEnded If you want to test this firmware, you can download it from here:
https://github.com/DragonBluep/openwrt/actions/runs/3340633201

It should now be ready for review.

Lucky1openwrt commented:
the image file size can not be above "IMAGE_SIZE := 6208k"
if the image file is bigger it will write over the caldata at 6600000

DragonBluep commented:
> the image file size can not be above "IMAGE_SIZE := 6208k" if the image file is bigger it will write over the caldata at 6600000

No, squashfs update will skip caldata partition. We use mtd-concat to handle it.

Lucky1openwrt commented:
but the 19.07 file you are sysupgrading from don't know that

DeadEnded commented:
### Sysupgrade
Okay - I first tested the sysupgrade.bin from the git build... everything looks good.

![image](https://user-images.githubusercontent.com/45110141/198687906-f585b0a6-0f74-43c8-9998-9fb631f49767.png)


If there are any specifics that need tested let me know... ethernet and wifi are up, system/kernel log just have one error:
```
Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.256818] ath9k_pci_owl_loader 0000:00:11.0: Direct firmware load for ath9k-eeprom-pci-0000:00:11.0.bin failed with error -2
Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.268269] ath9k_pci_owl_loader 0000:00:11.0: Falling back to sysfs fallback for: ath9k-eeprom-pci-0000:00:11.0.bin
Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.280164] ath9k_pci_owl_loader 0000:00:12.0: Direct firmware load for ath9k-eeprom-pci-0000:00:12.0.bin failed with error -2
Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.291610] ath9k_pci_owl_loader 0000:00:12.0: Falling back to sysfs fallback for: ath9k-eeprom-pci-0000:00:12.0.bin
```
I don't know how much of an issue this is... might be able to be ignored?

Next, I ran the sysupgrade.bin from a build using the default config from 22.03 on my branch (because I didn't know one would be built from git).  Not sure what differences there might be between the git build and mine (I believe mine is from master since that is my upstream)... so this might be a waste, but wanted to try and be thorough...

![image](https://user-images.githubusercontent.com/45110141/198689429-75b845d1-2901-4aa7-a766-df1498204ebd.png)

Everything looks good here as well - just the same errors as above in system/kernel logs.

### Factory.bin
Next, I flashed stock firmware `TEW673GRUA1_FW100B40.bin`.
Once this was complete, I used the stock GUI and uploaded the factory.bin from the git build.
Some undetermined amount of time later - Success!
I restored by config backup, and everything looks to be up and running.
Since my old build was somewhat dated, there are some new features I didn't recognize (dashboard, attended sysupgrade, etc.).
This will now be my daily driver - if there are any tests that need to be completed - please let me know.

Thank you once again for all your help!
I can confirm at a basic "It boots and looks right" level, both factory.bin and sysupgrade.bin work!

 - DeadEnd

DragonBluep commented:
> but the 19.07 file you are sysupgrading from don't know that

@Lucky1openwrt That's why we not support upgrade from 19.07.

> > Just confirming - you intentionally removed the `SUPPORTED_DEVICES += tew-673gru` from generic.mk?
> > I see some have it and some don't - but just wanted to confirm.
> 
> This is used to support the upgrade from the old version openwrt 19.07, but since we have changed the partition table, it is better not to be compatible with old version (but we can still apply it by force update).

Lucky1openwrt commented:
you will have to make a firmware file bigger then "IMAGE_SIZE := 6208k" 0x610000
and if the firmware allows you to sysupgrade to it 
it will start writing it to 0x050000 and go up 
if the file size is above 0x610000 it will continue writing over the cal data at 0x660000

DragonBluep commented:
mtd will write image to firmware partiton, not fwconcat0 partition. there is an HAL to handle it.

Lucky1openwrt commented:
I'm only coming from the dir-825
but as long as you have tested & put warnings
I have not read al the comments so my bad 
so I'm sure I have missed the not syupgrade form v19
it seems like a chicken and egg thing but as long as you have tested and confirmed 
it write around caldata it's all good

DragonBluep commented:
> ```
> Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.256818] ath9k_pci_owl_loader 0000:00:11.0: Direct firmware load for ath9k-eeprom-pci-0000:00:11.0.bin failed with error -2
> Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.268269] ath9k_pci_owl_loader 0000:00:11.0: Falling back to sysfs fallback for: ath9k-eeprom-pci-0000:00:11.0.bin
> Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.280164] ath9k_pci_owl_loader 0000:00:12.0: Direct firmware load for ath9k-eeprom-pci-0000:00:12.0.bin failed with error -2
> Fri Oct 28 09:49:19 2022 kern.warn kernel: [   40.291610] ath9k_pci_owl_loader 0000:00:12.0: Falling back to sysfs fallback for: ath9k-eeprom-pci-0000:00:12.0.bin
> ```
> 
> 
>     
>   
> 
> I don't know how much of an issue this is... might be able to be ignored?

I have no idea about it, but it should be okay, just some warnings and wifi is up. @chunkeey can you have a look, thanks.


@DeadEnded we need to merge final `art` partition into fwconcat1. From @Lucky1openwrt's explanation https://github.com/openwrt/openwrt/pull/4603#discussion_r1008261947, this partition was created by the previous openwrt v19.07 or older. At the same time, we also need to change `IMAGE_SIZE := 7808k`

DeadEnded commented:
Should be all set, image size changed, and last partition merged back together.
Thanks!

DeadEnded commented:
I just noticed / realized we changed the `wps` LED to `power`.
I checked and confirmed that the LED setup is actually for wps (whether that's used in openWRT or not).
There is a second LED for power.

So, I am going to change the naming back to `led_wps: wps` (unless there is a naming nomenclature that is preferred) and see if I can figure out if the power LED is tied to a gpio or if it's just hardwired to power.

hauke commented:
Thank you for your patch, I applied it to master in 12cee869890853716ff1ee2dbd0a89c87a0ee544.

DragonBluep commented:
@hauke Could you please cherry-pick it to v22.03 branch? Due to the limitation of the stock firmware, we only have 6144k space to generate `factory.bin`. I'm afraid the next stable version v23. x is not able to generate a factory image. There is about 250KiB free space left in the `factory.bin` with current master branch (include LuCI).

hauke commented:
The phy-mask attribute was removed in master in dae392752761128e13ede484217d806d643cbf2f. Is this still in openwrt 22.03?

DragonBluep commented:
Oh, my bad. We still keep phy-mask in 22.03 branch. But it seems useless? I'm not sure. c8c96b22c6ef1f0a6060d5a5f2a94ebdfc5b8c8c
In @DeadEnded's original commit, phy-mask = <0x10>;

Wirmos commented:
@DeadEnded , @DragonBluep Could you clarify why did you shrink image size to 'FACTORY_SIZE := 6144k' even fwconcat0 partition is 6208KiB?

DragonBluep commented:
@Wirmos The [stock firmware](https://downloads.trendnet.com/tew-673gru/firmware/tew-673gru_fw100b40.zip) only accepts images in this format: 6144k binary (0x600000) + signature. Openwrt can handle larger sysupgrade images.

Wirmos commented:
@DragonBluep Got it, thanks.

hauke commented:
I backported it to 22.03 too, see 1133a8f805a3e8d3a3a63269de0ba9a7df0340dd.
